grails-app/controllers/org/pih/warehouse/order/OrderController.groovy

```diff
@@ -1096,6 +1096,47 @@ class OrderController {
     }
 
     def orderSummary = {
+        Order order = Order.get(params.id)
+        render(template: "orderSummary", model: [orderInstance: order])
+    }
+
+    def itemStatus = {
+        Order order = Order.get(params.id)
+        render(template: "itemStatus", model: [orderInstance: order])
+    }
+
+    def itemDetails = {
+        Order order = Order.get(params.id)
+        render(template: "itemDetails", model: [orderInstance: order])
+    }
+
+    def orderAdjustments = {
+        Order order = Order.get(params.id)
+        render(template: "orderAdjustments", model: [orderInstance: order])
+    }
+
+    def orderShipments = {
+        Order order = Order.get(params.id)
+        render(template: "orderShipments", model: [orderInstance: order])
+    }
+
+    def orderInvoices = {
+        Order order = Order.get(params.id)
+        render(template: "orderInvoices", model: [orderInstance: order])
+    }
+
+    def orderDocuments = {
+        Order order = Order.get(params.id)
+        render(template: "orderDocuments", model: [orderInstance: order])
+    }
+
+    def orderComments = {
+        Order order = Order.get(params.id)
+        render(template: "orderComments", model: [orderInstance: order])
+    }
```

Comment:
I'm wondering if this is the result we wanted to have? 
I imagine that now, when we go to the PO view page, and go between the tabs, we everytime fetch the same order instance (with all the fields etc).

I know that probably anyway it is faster now, but maybe we should return to the view only fields that we really need, so maybe to return some kind of command/dto objects, e.g.
```groovy
class OrderCommentCommand {
  String orderId
  Comments comments
}
```
(pseudocode below, probably could be simplified/be more clean)
```groovy
def orderComments = {
        Order order = Order.get(params.id)
        OrderCommentCommand orderCommentCommand = 
             new OrderCommentCommand(orderId: order?.id, comments: order?.comments)
        render(template: "orderComments", model: [orderCommentCommand: orderCommentCommand])
    }
```
etc.

Reply:
You still fetch the same order in your example

Reply:
PS. Data services could be a good candidate if we decided to create custom queries for those, and there to use the `@Query` annotation.

Reply:
Initially, I wanted to pass the entire order instance object in params, but it was stringified, so I changed it to sending order's id only, and this additional `get` did not really affect the performance in this case since the rendering is a problematic part now. Plus adding a command for each tab adds more work to be done for this "quick fix"

Reply:
Yup, those are only suggestions that came to my mind, if we wanted to improve the performance even more, but I agree with you that it adds some scope which we might not want to add now.
Still, even though I gave those suggestions, I think they wouldn't be a game changer, because we fetch only one order there, so the differences shouldn't be significant versus a situation where we'd have to loop over 100 orders.

Reply:
I recorded a video to describe what I am seeing and what I think we need to do. 
https://watch.screencastify.com/v/Wm2n3lW41EfiqMpAxT0M

I hadn't read @kchelstowski's comments before the video but we're basically saying the same thing

> I know, but I don't return the whole order object to the view, but just the part, that we really need there (this is why I wrote "with all the fields etc")

So I agree with @kchelstowski on the problem but I'm not sure if we need separate command objects just yet. 

The performance problems with the ajax requests are probably all bottled up in the template GSPs. Some of those are likely causing tons of N+1 queries. A more thoughtful approach would be to rewrite each of those to gather just the data needed for the page, hopefully in a single query.

I imagine there would be at least two methods 

* Order getOrderSummary(String id)
* Order getOrderDetails(String id)

that queries the database (including joins) to gather all of the data needed for that particular tab.

If that's not enough we can add other methods as data gatherers for the other closures.

* Order getOrderItemStatuses(String id) 
* Order getOrderItemDetails(String id)
* etc

Hopefully, only one of these methods will be the expensive one but I'm hoping we're able to limit it use to just the tabs that require the expensive view-backed data (i.e. invoice or status data that is taking a long time to load).

Assuming we can get everything we need packed into the Order object (or a map) then we can leave the command object for later. 

For now, the important part is just trying to add some service methods that pull as much data as possible in one query. And honestly, I just need the summary tab to load in under 1s right now. Taking a little longer for subsequent tabs is fine until we have time to tackle the underlying slowness.